File: src/main/java/nl/crosscode/x509certbuddy/utils/EditorUtils.java
package nl.crosscode.x509certbuddy.utils;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.Inlay;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.wm.ToolWindow;
import com.intellij.openapi.wm.ToolWindowManager;
import nl.crosscode.x509certbuddy.decoder.CertRetriever;
import nl.crosscode.x509certbuddy.decoder.RetrievedCert;
import nl.crosscode.x509certbuddy.ui.CertEditorElementRender;
import nl.crosscode.x509certbuddy.x509CertAssistantFactory;

import java.security.cert.CertificateEncodingException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.*;
import java.util.stream.Collectors;

public class EditorUtils {

    private static Dictionary<Editor,List<Inlay>> inlays = new Hashtable<>();
    private static Dictionary<Editor,List<X509Certificate>> certs = new Hashtable<>();
    public void readCertsFromEditor(Editor editor) {
        if (editor==null) return;
        Project project = editor.getProject();
        String allText = editor.getDocument().getText();
        try {
            CertRetriever certRetriever = new CertRetriever(editor);
            List<RetrievedCert> certs = certRetriever.retrieveCerts(allText);
            if (project!=null) {
                ToolWindow tw = ToolWindowManager.getInstance(project).getToolWindow("X.509 Cert Buddy");
                if (tw != null) {
                    x509CertAssistantFactory.getInstance(project).addCerts(certs.stream().map(RetrievedCert::getCertificate).collect(Collectors.toList()));
                    tw.show();
                }
            }
            addCertsToEditor(certs, editor);
        } catch (CertificateException ex) {}
    }

    public synchronized void removeEditor(Editor editor) {
        certs.remove(editor);
        List<Inlay> inlayList = inlays.get(editor);
        if (inlayList!=null) {
            inlayList.stream().forEach(x->x.dispose());
        }
        inlays.remove(editor);
    }

    private synchronized void addCertsToEditor(List<RetrievedCert> retrievedCerts, Editor editor) {
        List<X509Certificate> certsOfEditor = certs.get(editor);
        boolean certsUpdate = true;
        if (certsOfEditor!=null&&certsOfEditor.size()==retrievedCerts.size()) {
            certsUpdate = false;
            for (RetrievedCert cert : retrievedCerts) {
                if (certsOfEditor.stream().noneMatch(x-> {
                    try {
                        return Arrays.equals(x.getEncoded(),cert.getCertificate().getEncoded());
                    } catch (CertificateEncodingException e) {
                        throw new RuntimeException(e);
                    }
                })) {
                    certsUpdate = true;
                    break;
                }
            }
        }
        if (!certsUpdate) return;
        removeEditor(editor);
        List<Inlay> inlayList = new ArrayList<>();
        inlays.put(editor,inlayList);
        certsOfEditor = new ArrayList<>();
        certs.put(editor,certsOfEditor);
        for (RetrievedCert retrievedCert : retrievedCerts) {
            certsOfEditor.add(retrievedCert.getCertificate());
            inlayList.add(editor.getInlayModel().addBlockElement(retrievedCert.getOffset(),false,true,0,new CertEditorElementRender(retrievedCert.getCertificate())));
        }
        editor.getComponent().repaint();
    }
}

File: src/main/kotlin/nl/crosscode/x509certbuddy/utils/EditorUtils.kt
package nl.crosscode.x509certbuddy.utils

import com.intellij.openapi.application.Application
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.ReadAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.Inlay
import com.intellij.openapi.wm.ToolWindowManager
import com.intellij.util.application
import nl.crosscode.x509certbuddy.decoder.CertRetriever
import nl.crosscode.x509certbuddy.decoder.RetrievedCert
import nl.crosscode.x509certbuddy.ui.CertEditorElementRender
import nl.crosscode.x509certbuddy.X509CertAssistantFactory
import java.security.cert.CertificateEncodingException
import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import java.util.*
import java.util.function.Consumer
import java.util.stream.Collectors

class EditorUtils {
    fun readCertsFromEditor(editor: Editor?) {
        if (editor == null) return
        val project = editor.project
        val allText = editor.document.text
        try {
            var certs: List<RetrievedCert>? = null
            application.runReadAction {
                val certRetriever = CertRetriever(editor)
                certs = certRetriever.retrieveCerts(allText)
            }
            if (project != null && certs != null) {
                val tw = ToolWindowManager.getInstance(project).getToolWindow("X.509 Cert Buddy")
                if (tw != null) {
                    tw.show()
                    application.runWriteAction {
                        X509CertAssistantFactory.getInstance(project)
                            ?.addCerts(certs!!.map { it.certificate } )
                    }

                }
            }
            if (certs!=null) {
                application.runWriteAction {
                    addCertsToEditor(certs!!, editor)
                }
            }
        } catch (ignored: CertificateException) {
        }
    }

    private fun removeEditor(editor: Editor?) {
        synchronized(lock) {
            certs.remove(editor)
            val inlayList = inlays[editor]
            inlayList?.forEach(Consumer { obj: Inlay<*>? -> obj!!.dispose() })
            inlays.remove(editor)
        }
    }

    private fun addCertsToEditor(retrievedCerts: List<RetrievedCert>, editor: Editor) {
        synchronized(lock) {
            var certsOfEditor = certs[editor]
            var certsUpdate = true
            application.runReadAction {
                if (certsOfEditor != null && certsOfEditor.size == retrievedCerts.size) {
                    certsUpdate = false
                    retrievedCerts.any {rc ->
                        certsOfEditor.any { rc.certificate==it }
                    }.let { if (it) {
                        certsUpdate = true
                    }}
                }
            }
            if (!certsUpdate) return
            application.runWriteAction {
                removeEditor(editor)
                val inlayList: MutableList<Inlay<*>?> = ArrayList()
                inlays.put(editor, inlayList)
                certsOfEditor = ArrayList()
                certs.put(editor, certsOfEditor)
                for (retrievedCert in retrievedCerts) {
                    certsOfEditor.add(retrievedCert.certificate)
                    inlayList.add(
                        editor.inlayModel.addBlockElement(
                            retrievedCert.offset,
                            false,
                            true,
                            0,
                            CertEditorElementRender(retrievedCert.certificate)
                        )
                    )
                }
            }
        }
        editor.component.repaint()
    }

    companion object {
        private val lock = Any()
        private val inlays: Dictionary<Editor, List<Inlay<*>?>> = Hashtable()
        private val certs: Dictionary<Editor, MutableList<X509Certificate>> = Hashtable()
    }
}
